src/price_providers/dune_pricing.py

```diff
@@ -0,0 +1,56 @@
+from typing import Optional
+import dotenv, os
+from dune_client.types import QueryParameter
+from dune_client.client import DuneClient
+from dune_client.query import QueryBase
+from src.helpers.config import get_web3_instance, get_logger
+from src.constants import FETCH_PRICE_QUERY_ID, QUERY_BUFFER_TIME
+
+dotenv.load_dotenv()
+dune_api_key = os.getenv("DUNE_API_KEY")
+dune = DuneClient.from_env()
+
+
+class DunePriceProvider:
+    """
+    Purpose of this class is to fetch historical token prices from Dune.
+    """
+
+    def __init__(self) -> None:
+        self.web3 = get_web3_instance()
+        self.logger = get_logger()
+
+    def get_price(self, block_number: int, token_address: str) -> Optional[float]:
+        """
+        Function returns Dune price for a token address,
+        closest to and at least as large as the block timestamp for a given tx hash.
+        """
+        try:
+            start_timestamp = self.web3.eth.get_block(block_number)["timestamp"]
```

Comment:
Would it make sense to change this to `.get("timestamp")` and manually log something if the result is `None`? I think this is the only part that could throw a KeyError.